api/agent/static_pool.go

```diff
@@ -57,77 +52,64 @@ func (rp *staticRunnerPool) Runners(call pool.RunnerCall) ([]pool.Runner, error)
 }
 
 func (rp *staticRunnerPool) AddRunner(address string) error {
-	rp.rMtx.Lock()
-	defer rp.rMtx.Unlock()
-
 	r, err := rp.generator(address, rp.runnerCN, rp.pki)
 	if err != nil {
 		logrus.WithField("runner_addr", address).Warn("Failed to add runner")
 		return err
 	}
-	rp.runners = append(rp.runners, r)
+
+	rp.rMtx.Lock()
+
+	isFound := false
+	for _, r := range rp.runners {
+		if r.Address() == address {
+			isFound = true
+			break
+		}
+	}
+	if !isFound {
+		rp.runners = append(rp.runners, r)
+	}
+
+	rp.rMtx.Unlock()
 	return nil
 }
 
 func (rp *staticRunnerPool) RemoveRunner(address string) {
-	rp.rMtx.Lock()
-	defer rp.rMtx.Unlock()
 
-	ctx, cancel := context.WithTimeout(context.Background(), staticPoolShutdownTimeout)
-	defer cancel()
+	var toRemove pool.Runner
+
+	rp.rMtx.Lock()
 
 	for i, r := range rp.runners {
 		if r.Address() == address {
-			err := r.Close(ctx)
-			if err != nil {
-				logrus.WithError(err).WithField("runner_addr", r.Address()).Error("Failed to close runner")
-			}
-			// delete runner from list
+			toRemove = r
 			rp.runners = append(rp.runners[:i], rp.runners[i+1:]...)
-			return
+			break
 		}
 	}
+
+	rp.rMtx.Unlock()
+
+	if toRemove != nil {
+		toRemove.Close(context.Background())
+	}
 }
 
-// Shutdown blocks waiting for all runners to close, or until ctx is done
+// Shutdown blocks waiting for all runners to close
 func (rp *staticRunnerPool) Shutdown(ctx context.Context) (e error) {
+
+	var toRemove []pool.Runner
+
 	rp.rMtx.Lock()
-	defer rp.rMtx.Unlock()
 
-	ctx, cancel := context.WithTimeout(context.Background(), staticPoolShutdownTimeout)
-	defer cancel()
+	toRemove = rp.runners[:]
```

Comment:
nit: could `toRemove :=` here and avoid the 3 lines of var decl, no behavior change.